cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,304 @@
+		kubeClient: kubeClient,
+		client:     client,
+		tags:       tags,
+		fwManager:  fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			fc.fwCache.mu.RLock()
+			isEqual := cmp.Equal(currentFirewall, fc.fwCache.firewall)
+			fc.fwCache.mu.RUnlock()
+
+			if isEqual {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, firewallReconcileRequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if fm.fwCache.firewall != nil {
+		fm.fwCache.mu.RLock()
+		fwID := fm.fwCache.firewall.ID
+		fm.fwCache.mu.RUnlock()
+
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fwID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return cmp.Equal(fw.Name, fm.workerFirewallName)
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	fm.fwCache.mu.RUnlock()
+
+	fwSvc := fm.client.Firewalls
+
+	fm.fwCache.mu.RLock()
+	isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
+	fm.fwCache.mu.RUnlock()
```

Comment:
The two synchronized sections could be combined into one, avoiding an unlock/lock, and I would also do the return before assigning `fwSvc`, just for readability (since it's not used in the early return case):

```go
fm.fwCache.mu.RLock()
targetFirewall := fm.fwCache.firewall
isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
fm.fwCache.mu.RUnlock()

if isEqual {
    return nil
}

fwSvc := fm.client.Firewalls
...
```

Reply:
Ahhh yes. Much cleaner!